core/schema/service.go

```diff
@@ -278,3 +286,74 @@ func getIDforSchema(ns, schema, dataUUID string) string {
 	key := fmt.Sprintf("%s-%s-%s", ns, schema, dataUUID)
 	return uuid.NewSHA1(uuid.NameSpaceOID, []byte(key)).String()
 }
+
+func (s *Service) DetectSchemaChange(namespace string, schemaName string, fromVersion string, toVersion string, depth string) (*stencilv1beta1.SchemaChangedEvent, error) {
+	ctx := context.Background()
+	if fromVersion == "" || toVersion == "" {
+		fmt.Printf("schema service name %s", schemaName)
+		latestVersion, err := s.repo.GetLatestVersion(ctx, namespace, schemaName)
+		if err != nil {
+			return nil, fmt.Errorf("error getting latest version - %s", err.Error())
+		}
+		if latestVersion == 1 {
+			return nil, fmt.Errorf("only one version exists for schema - %s", schemaName)
+		}
+		if fromVersion == "" {
+			fromVersion = strconv.Itoa(int(latestVersion - 1))
+		}
+		if toVersion == "" {
+			toVersion = strconv.Itoa(int(latestVersion))
+		}
+	}
+	if depth == "" {
+		depth = strconv.Itoa(-1)
```

Comment:
`func (s *Service) DetectSchemaChange(namespace string, schemaName string, fromVersion string, toVersion string, depth string) (*stencilv1beta1.SchemaChangedEvent, error) {
	ctx := context.Background()
	fmt.Printf("fromVersion in request - %s", fromVersion)
	fmt.Printf("toVersion in request- %s", toVersion)
	fromVer, toVer, err := s.getVersions(ctx, namespace, schemaName, fromVersion, toVersion)
	if err != nil {
		return nil, err
	}
	fmt.Printf("fromVersion after conversion - %d", fromVer)
	fmt.Printf("toVersion after conversion- %d", toVer)
	if depth == "" {
		depth = strconv.Itoa(-1)
	}
	depth64, errDepth := strconv.ParseInt(depth, 10, 32)
	if errDepth != nil {
		return nil, fmt.Errorf("invalid depth")
	}
	_, fromVerData, fromVerDataError := s.Get(ctx, namespace, schemaName, int32(fromVer))
	if fromVerDataError != nil {
		return nil, fmt.Errorf("error getting data for version %d - %s", fromVer, fromVerDataError.Error())
	}
	_, toVerData, toVerDataError := s.Get(ctx, namespace, schemaName, int32(toVer))
	if toVerDataError != nil {
		return nil, fmt.Errorf("error getting data for version %d - %s", toVer, toVerDataError.Error())
	}
	req := &changedetector.ChangeRequest{
		NamespaceID: namespace,
		SchemaName:  schemaName,
		OldData:     fromVerData,
		NewData:     toVerData,
		Version:     int32(toVer),
		Depth:       int32(depth64),
	}
	sce, err := s.changeDetectorService.IdentifySchemaChange(ctx, req)
	if err != nil {
		return sce, fmt.Errorf("got error while identifying schema change for namespace : %s, schema: %s, version: %d, %s", req.NamespaceID, req.SchemaName, req.Version, err.Error())
	}
	if err := s.sendNotification(ctx, sce, req); err != nil {
		return sce, err
	}
	return sce, nil
}

func (s *Service) getVersions(ctx context.Context, namespace, schemaName, fromVersion, toVersion string) (int, int, error) {
	var toVer int
	var fromVer int
	if toVersion == "" {
		toVer, err := s.repo.GetLatestVersion(ctx, namespace, schemaName)
		if err != nil {
			return 0, 0, fmt.Errorf("error getting latest version - %s", err.Error())
		}
		if toVer == 1 {
			return 0, 0, fmt.Errorf("only one version exists for schema - %s", schemaName)
		}
	}
	toVer, err := strconv.Atoi(toVersion)
	if err != nil {
		return 0, 0, fmt.Errorf("invalid toVersion format")
	}
	if fromVersion == "" {
		fromVer = toVer - 1
	}
	fromVer, err = strconv.Atoi(fromVersion)
	if err != nil {
		return 0, 0, fmt.Errorf("invalid fromVersion format")
	}

	if fromVer <= 0 || toVer <= 0 {
		return 0, 0, fmt.Errorf("from/to Version should be greater than equal to 1")
	}

	if fromVer >= toVer {
		return 0, 0, fmt.Errorf("'from' should be less than 'to'")
	}
	return fromVer, toVer, nil
}
`

Just take a look at the above method.
Have extracted out the logic for versioning in getVersions method. Also, added a condition from <=0 for versions and little changes in if conditions.

Reply:
In this method, if toVersion is an empty string, the value of toVer determined by s.repo.GetLatestVersion will be overwritten by zero because it immediately attempts to convert the empty toVersion string to an integer, which results in an error?

Reply:
if its empty then we need to call db right?